File: lib/models/money.dart
/// Phase 6 — multi-currency framework on the Flutter side. Mirrors
/// `backend/src/lib/money.js`. UZS is the only active currency at launch;
/// KZT/KGS activate in Phase 7.
class Money {
  final num amount;
  final String currency;

  const Money(this.amount, [this.currency = 'UZS']);

  /// Parse from any of:
  /// - `{amount, currency, formatted?}` (Phase 6 backend canonical shape)
  /// - bare number (legacy — assumes UZS)
  factory Money.fromJson(dynamic raw, {String fallbackCurrency = 'UZS'}) {
    if (raw == null) return Money(0, fallbackCurrency);
    if (raw is num) return Money(raw, fallbackCurrency);
    if (raw is Map) {
      final a = raw['amount'];
      final c = (raw['currency'] ?? fallbackCurrency).toString();
      return Money(a is num ? a : num.tryParse(a?.toString() ?? '0') ?? 0, c);
    }
    return Money(num.tryParse(raw.toString()) ?? 0, fallbackCurrency);
  }

  Map<String, dynamic> toJson() => {
        'amount': amount,
        'currency': currency,
      };

  static const _meta = <String, _CurrencyMeta>{
    'UZS': _CurrencyMeta("so'm", 0),
    'KZT': _CurrencyMeta('₸', 0),
    'KGS': _CurrencyMeta('сом', 0),
    'RUB': _CurrencyMeta('₽', 2),
    'USD': _CurrencyMeta(r'$', 2),
  };

  /// Locale-aware string. `locale` is one of 'uz' | 'ru' | 'en' | 'kk'.
  /// Returns e.g. "45 000 so'm" / "45,000 \$" / "1 234.56 ₽".
  String format([String locale = 'ru']) {
    final meta = _meta[currency] ?? _meta['UZS']!;
    final fixed = amount.toStringAsFixed(meta.decimals);
    final parts = fixed.split('.');
    final intPart = parts[0];
    final frac = parts.length > 1 ? parts[1] : null;
    final sep = locale.startsWith('en') ? ',' : ' ';
    final grouped = intPart.replaceAllMapped(
      RegExp(r'\B(?=(\d{3})+(?!\d))'),
      (_) => sep,
    );
    final numStr = frac != null ? '$grouped.$frac' : grouped;
    return '$numStr ${meta.symbol}';
  }

  Money operator +(Money other) {
    if (other.currency != currency) {
      throw StateError('Cannot add $currency and ${other.currency}');
    }
    return Money(amount + other.amount, currency);
  }

  Money operator -(Money other) {
    if (other.currency != currency) {
      throw StateError('Cannot subtract $currency and ${other.currency}');
    }
    return Money(amount - other.amount, currency);
  }

  bool get isZero => amount == 0;
}

class _CurrencyMeta {
  final String symbol;
  final int decimals;
  const _CurrencyMeta(this.symbol, this.decimals);
}

File: backend/src/lib/money.js
// Phase 6 — multi-currency framework. UZS is the only active currency at
// launch; KZT/KGS activate in Phase 7. Wraps numeric amounts with their
// currency so serialization carries both fields uniformly.

// Currencies we care about across CIS launch markets.
const CURRENCIES = {
  UZS: { code: 'UZS', symbol: "so'm", decimals: 0 },
  KZT: { code: 'KZT', symbol: '₸', decimals: 0 },
  KGS: { code: 'KGS', symbol: 'сом', decimals: 0 },
  RUB: { code: 'RUB', symbol: '₽', decimals: 2 },
  USD: { code: 'USD', symbol: '$', decimals: 2 },
};

function isSupported(code) {
  return Object.prototype.hasOwnProperty.call(CURRENCIES, code);
}

// Build a Money value object. Negative amounts are allowed (refunds, deltas).
function money(amount, currency = 'UZS') {
  const code = String(currency || 'UZS').toUpperCase();
  if (!isSupported(code)) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  const n = Number(amount);
  if (!Number.isFinite(n)) {
    throw new Error(`Invalid money amount: ${amount}`);
  }
  return { amount: n, currency: code };
}

// Serialize for API responses. Returns `{ amount, currency, formatted }`
// so clients can either render `formatted` directly or apply their own
// locale-aware formatting.
function toJson(m, locale = 'ru') {
  if (m == null) return null;
  return {
    amount: m.amount,
    currency: m.currency,
    formatted: format(m, locale),
  };
}

// Locale-aware string. Currency symbols come from CURRENCIES; thousand
// separators follow the locale convention.
function format(m, locale = 'ru') {
  if (m == null) return '';
  const meta = CURRENCIES[m.currency] || CURRENCIES.UZS;
  const fixed = m.amount.toFixed(meta.decimals);
  const [intPart, fracPart] = fixed.split('.');
  const sep = (locale || 'ru').startsWith('en') ? ',' : ' ';
  const grouped = intPart.replace(/\B(?=(\d{3})+(?!\d))/g, sep);
  const numStr = fracPart != null ? `${grouped}.${fracPart}` : grouped;
  return `${numStr} ${meta.symbol}`;
}

// Convert between currencies using an explicit rates table.
// rates: { from: { to: rate } }, e.g. { UZS: { KZT: 0.038 } }.
function convert(m, toCurrency, rates) {
  if (m == null) return null;
  const target = String(toCurrency).toUpperCase();
  if (m.currency === target) return m;
  const rate = rates && rates[m.currency] && rates[m.currency][target];
  if (!Number.isFinite(rate)) {
    throw new Error(`No conversion rate for ${m.currency}→${target}`);
  }
  return money(m.amount * rate, target);
}

module.exports = {
  money,
  toJson,
  format,
  convert,
  isSupported,
  CURRENCIES,
};
